# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [ ]:
import pandas as pd
import numpy as np

## Input Dataset

In [ ]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [ ]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [ ]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df[['Geography', 'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary', 'Exited']].copy()


## Preprocessing

In [ ]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df = pd.get_dummies(df, columns=['Geography', 'Gender'], drop_first=True)


In [ ]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
X = df.drop('Exited', axis=1)  # Mengambil semua kolom kecuali 'Exited' sebagai fitur
y = df['Exited']  # Mengambil kolom 'Exited' sebagai target


In [ ]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

#START CODE
# Inisialisasi scaler
scaler = MinMaxScaler()

# Menggunakan scaler untuk melakukan normalisasi pada fitur-fitur
X_transform = scaler.fit_transform(X)

In [ ]:
#Ini digunakan jika dilakukan scaler/Normalisas. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)
#X_transform = X.copy()

## Train-Test Split

In [ ]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

## Model1

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV


#START CODE
model1 = RandomForestClassifier()
params = {
    'n_estimators': [100, 200, 300],  # Jumlah pohon dalam forest
    'max_depth': [None, 10, 20, 30],  # Kedalaman maksimum setiap pohon
    'min_samples_split': [2, 5, 10],  # Jumlah sampel minimum yang dibutuhkan untuk membagi node
    'min_samples_leaf': [1, 2, 4]  # Jumlah sampel minimum yang dibutuhkan di setiap leaf node
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model1,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

{'max_depth': 10,
 'min_samples_leaf': 2,
 'min_samples_split': 5,
 'n_estimators': 200}

In [ ]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix

# Melakukan prediksi menggunakan model terbaik
y_pred = grid.predict(X_test)

# Menampilkan classification report
print("Classification Report:\n", classification_report(y_test, y_pred))

# Menampilkan confusion matrix
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))

# Menampilkan akurasi
accuracy = accuracy_score(y_test, y_pred)
print("\nAccuracy Score:", accuracy)


Classification Report:
               precision    recall  f1-score   support

           0       0.87      0.97      0.92      1983
           1       0.80      0.46      0.58       517

    accuracy                           0.86      2500
   macro avg       0.83      0.71      0.75      2500
weighted avg       0.86      0.86      0.85      2500


Confusion Matrix:
 [[1923   60]
 [ 281  236]]

Accuracy Score: 0.8636


## Model2
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
# Inisialisasi model SVM
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

model2 = SVC()

# Definisikan hyperparameter yang akan dioptimasi
params = {
    'C': [0.1, 1, 10],  # Parameter penalti
    'kernel': ['linear', 'rbf', 'poly'],  # Jenis kernel
    'gamma': ['scale', 'auto', 0.1, 1],  # Koefisien kernel
}

# Inisialisasi GridSearchCV
grid = GridSearchCV(
    estimator=model2,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,  # Jumlah core CPU yang digunakan
    cv=10  # 10-fold cross validation
)
#END CODE

# Melakukan grid search untuk mencari hyperparameter terbaik
grid.fit(X_train, y_train)

# Menampilkan hyperparameter terbaik
grid.best_params_

{'C': 10, 'gamma': 1, 'kernel': 'poly'}

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix

# Melakukan prediksi menggunakan model terbaik (hasil dari GridSearchCV)
y_pred = grid.predict(X_test)

# Menampilkan classification report
print("Classification Report:\n", classification_report(y_test, y_pred))

# Menampilkan confusion matrix
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))

# Menampilkan akurasi
accuracy = accuracy_score(y_test, y_pred)
print("\nAccuracy Score:", accuracy)


              precision    recall  f1-score   support

           0       0.87      0.98      0.92      1983
           1       0.82      0.43      0.56       517

    accuracy                           0.86      2500
   macro avg       0.85      0.70      0.74      2500
weighted avg       0.86      0.86      0.84      2500


[[1936   47]
 [ 297  220]]

0.8624


## Model3
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

# Inisialisasi model K-NN
model3 = KNeighborsClassifier()

# Definisikan hyperparameter yang akan dioptimasi
params = {
    'n_neighbors': [3, 5, 7],  # Jumlah tetangga terdekat
    'weights': ['uniform', 'distance'],  # Metode bobot
    'p': [1, 2],  # Metrik jarak (1 untuk Manhattan, 2 untuk Euclidean)
}

# Inisialisasi GridSearchCV
grid = GridSearchCV(
    estimator=model3,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,  # Jumlah core CPU yang digunakan
    cv=10  # 10-fold cross validation
)
#END CODE


# Melakukan grid search untuk mencari hyperparameter terbaik
grid.fit(X_train, y_train)

# Menampilkan hyperparameter terbaik
grid.best_params_

{'n_neighbors': 7, 'p': 1, 'weights': 'distance'}

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix

# Melakukan prediksi menggunakan model terbaik (hasil dari GridSearchCV)
y_pred = grid.predict(X_test)

# Menampilkan classification report
print("Classification Report:\n", classification_report(y_test, y_pred))

# Menampilkan confusion matrix
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))

# Menampilkan akurasi
accuracy = accuracy_score(y_test, y_pred)
print("\nAccuracy Score:", accuracy)


Classification Report:
               precision    recall  f1-score   support

           0       0.85      0.95      0.89      1983
           1       0.63      0.34      0.45       517

    accuracy                           0.82      2500
   macro avg       0.74      0.65      0.67      2500
weighted avg       0.80      0.82      0.80      2500


Confusion Matrix:
 [[1879  104]
 [ 339  178]]

Accuracy Score: 0.8228


## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

Dari 3 Metode yang telah saya buat terlihat model yang terbaik merupakan model RandomForestClassifier, hal ini dikarenakan metode RandomForestClassifier memiliki accuracy paling tinggi yaitu dengan nilai Accuracy Score: 0.8636. Sedangkan pada model SVM 0.8624 dan KNeighborsClassifier dengan nilai 0.8228